sw/device/lib/crypto/drivers/keymgr.c

```diff
@@ -0,0 +1,257 @@
+  if (launder32(status) == KEYMGR_OP_STATUS_STATUS_VALUE_IDLE) {
+    HARDENED_CHECK_EQ(status, KEYMGR_OP_STATUS_STATUS_VALUE_IDLE);
+    return OTCRYPTO_OK;
+  }
+  return OTCRYPTO_RECOV_ERR;
+}
+
+/**
+ * Set diversification input and start the key manager operation.
+ *
+ * Ensure the key manager is idle before calling this function.
+ *
+ * @param diversification Diversification input for the key derivation.
+ */
+static void keymgr_start(keymgr_diversification_t diversification) {
+  // Set the version.
+  abs_mmio_write32(kBaseAddr + KEYMGR_KEY_VERSION_REG_OFFSET,
+                   diversification.version);
+  // Set the salt.
+  for (size_t i = 0; i < kKeymgrSaltNumWords; i++) {
+    abs_mmio_write32(
+        kBaseAddr + KEYMGR_SALT_0_REG_OFFSET + (i * sizeof(uint32_t)),
+        diversification.salt[i]);
+  }
+
+  // Issue the start command.
+  abs_mmio_write32(kBaseAddr + KEYMGR_START_REG_OFFSET,
+                   1 << KEYMGR_START_EN_BIT);
+}
+
+/**
+ * Wait for the key manager to finish an operation.
+ *
+ * Polls the key manager until it is no longer busy. If the operation completed
+ * successfully or the key manager was already idle, returns OTCRYPTO_OK. If
+ * there was an error during the operation, reads and clears the error code
+ * and returns OTCRYPTO_RECOV_ERR; the operation can be retried afterwards.
+ *
+ * @return OK or error.
+ */
+OT_WARN_UNUSED_RESULT
+static status_t keymgr_wait_until_done(void) {
+  // Poll the OP_STATUS register until it is something other than "WIP".
+  uint32_t reg;
+  uint32_t status;
+  do {
+    reg = abs_mmio_read32(kBaseAddr + KEYMGR_OP_STATUS_REG_OFFSET);
+    status = bitfield_field32_read(reg, KEYMGR_OP_STATUS_STATUS_FIELD);
+  } while (status == KEYMGR_OP_STATUS_STATUS_VALUE_WIP);
+
+  // Clear OP_STATUS by writing back the value we read.
+  abs_mmio_write32(kBaseAddr + KEYMGR_OP_STATUS_REG_OFFSET, reg);
+
+  // Check if the key manager reported errors. If it is already idle or
+  // completed an operation successfully, return an OK status.
+  switch (status) {
+    case KEYMGR_OP_STATUS_STATUS_VALUE_IDLE:
+      return OTCRYPTO_OK;
+    case KEYMGR_OP_STATUS_STATUS_VALUE_DONE_SUCCESS:
+      return OTCRYPTO_OK;
+    case KEYMGR_OP_STATUS_STATUS_VALUE_DONE_ERROR: {
+      // Clear the ERR_CODE register before returning.
+      uint32_t err_code =
+          abs_mmio_read32(kBaseAddr + KEYMGR_ERR_CODE_REG_OFFSET);
+      abs_mmio_write32(kBaseAddr + KEYMGR_ERR_CODE_REG_OFFSET, err_code);
+      return OTCRYPTO_RECOV_ERR;
+    }
+    case KEYMGR_OP_STATUS_STATUS_VALUE_WIP:
+      // Should be unreachable.
+      HARDENED_TRAP();
+      return OTCRYPTO_FATAL_ERR;
+    default:
+      // Should be unreachable.
+      HARDENED_TRAP();
+      return OTCRYPTO_FATAL_ERR;
+  }
+
+  // Should be unreachable.
+  HARDENED_TRAP();
+  return OTCRYPTO_FATAL_ERR;
+}
+
+/**
+ * Set the control register of the key manager.
+ *
+ * The CDI select bit is always set to false for this driver.
```

Comment:
nit: for documentation purposes, we should list this as SEALING_CDI. I assume that this change does not include support for attestation flows.